### PR TITLE
#315: Proper implementation of par template (with recursion)

### PR DESCRIPTION
**Pull request by paulodamaso**
For #315, made proper implementation of par template (with recursion) 
Performed changes:
- Corrected `par` template in `templates.xsl` for using recursion and correctly replacing paragraphs numbers;
- Added comment to `par.xsl`, corrected test description and set `ignore` to `false`.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#359](https://github.com/zerocracy/datum/issues/359) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#359](https://github.com/zerocracy/datum/issues/359) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/1802023302); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by amihaiemil** [comment]
@paulodamaso I'll have a look in 1-2 days, I'm away from my laptop for the weekend :(

File: xsl/templates.xsl [hunk 1]
```diff
@@ -184,8 +184,18 @@ SOFTWARE.
     </code>
   </xsl:template>
   <xsl:template name="par">
-    <xsl:param name="text"/>
-    <xsl:value-of select="$text"/>
+    <xsl:param name="text" />
+    <xsl:analyze-string select="$text" regex="(§[0-9]*)">
+      <xsl:matching-substring>
+        <xsl:variable select="regex-group(1)" name="paragraph" />
+        <a href="{concat('http://www.zerocracy.com/policy.html#', substring-after($paragraph,'§'))}">
```
Comment (amihaiemil) [inline xsl/templates.xsl]: @paulodamaso We should use ``https://`` here, I see it is enabled for zerocracy.com, but redirection from ``http://`` to ``https://`` is not :(

File: xsl-test/templates/project.xsl [hunk 2]
```diff
@@ -23,8 +23,9 @@ SOFTWARE.
       <xsl:with-param name="message" select="'Fails to format project name'"/>
       <xsl:with-param name="expected">
         <span>
-          <code>
-            <a href="https://www.0crat.com/p/ABCDEFGHI" title="Project ABCDEFGHI">
+          <code style="background-color:#daf1e0;">
```
Comment (amihaiemil) [inline xsl-test/templates/project.xsl]: @paulodamaso Why the ``background-color`` attribute, what does it do? And is the reindentation of ``<a>`` needed?

File: xsl-test/assertions.xsl [hunk 3]
```diff
@@ -15,21 +15,26 @@ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
...
           <xsl:value-of select="$message"/>
           <xsl:text> (actual "</xsl:text>
-          <xsl:value-of select="$actual_xml"/>
+          <xsl:copy-of select="$actual"/>
           <xsl:text>" is not equal to expected "</xsl:text>
-          <xsl:value-of select="$expected_xml"/>
+          <xsl:copy-of select="$expected"/>
           <xsl:text>")</xsl:text>
         </xsl:message>
       </xsl:if>
```
Comment (amihaiemil) [inline xsl-test/assertions.xsl]: @paulodamaso I don't think the changes to this file belong in this PR, or?

File: xsl-test/templates/project.xsl (hunk 2 above)
Comment (paulodamaso) [reply xsl-test/templates/project.xsl]: @amihaiemil This bug correction depends on https://github.com/zerocracy/datum/issues/339; I've had to put it here so the build doesn't fali.

File: xsl-test/assertions.xsl (hunk 3 above)
Comment (paulodamaso) [reply xsl-test/assertions.xsl]: @amihaiemil Same as above.

**Comment by paulodamaso** [comment]
@amihaiemil Thanks for the comments, made some changes and answered question, please take a look.

**Comment by paulodamaso** [comment]
@amihaiemil Let's do this the right way. I've removed all changes unrelated to #315, but now the build fails because some wrong code in master.

**Comment by amihaiemil** [comment]
@paulodamaso yes, the most correct would be to put this task on hold before #339 is solved. If you can merge the PR for #339 soon, then we can wait with this PR.

But if you think it will take longer, then we should close this PR (you keep your branch), and make another one later after the impediment is merged. This is so I don't lose this task, because I can't put PRs on hold ;(

**Comment by paulodamaso** [comment]
@amihaiemil Don't worry, I'll try to fix  #339 build problems as soon as possible. I think we will have enough time to this PR.

**Comment by amihaiemil** [comment]
@paulodamaso sure, thanks :D

**Comment by amihaiemil** [comment]
@paulodamaso I think you merged the waiting fix, right? But now the CI is still failing :D

**Comment by paulodamaso** [comment]
@amihaiemil I'm having some problems with the build, seems like some lib we are using does not support XSLT 2.0 `analyze-string` tag. The tests runs locally in my IDE, using Saxon 9.8.08 as processor, I'm suspecting that is a `Nokogiri` problem. Fortunately it does not work either with `Nokogiri` 1.8, so it was not caused by the downgrade of `Nokogiri` in #357 .

**Comment by paulodamaso** [comment]
@g4s8 This is a no-go here. `Nokogiri` does not support XPath 2.0 and XSLT 2.0 (see https://stackoverflow.com/questions/2838320/nokogiri-ruby-and-xpath and https://www.rubydoc.info/github/sparklemotion/nokogiri) because it uses `libxslt` which is an very very old lib. I don't think there is an elegant way of doing this recursive search using XSL 1.0, compared to `analyze-string` usage; so, I think we should try another way of building these files instead using `Nokogiri`, WDYT?

**Comment by amihaiemil** [comment]
@paulodamaso @g4s8 I honestly have no idea, but this seems like a bug. I would say: file a bug so you are paid for this research, put your task on hold (this one) until that one is fixed and you have all the necessary tools to remake this PR later. WDYT?

**Comment by g4s8** [comment]
@paulodamaso @amihaiemil right, I think we have to use Saxon to transform `xsl` (like it transformed to `html` in `Rakefile`), @paulodamaso please submit a bug to run xsl tests with Saxon instead of Nokogiri.

**Comment by paulodamaso** [comment]
@g4s8 Thanks for the direction, created https://github.com/zerocracy/datum/issues/365 for resolving the issue. Will correct this PR to ignore the test while the new bug is not corrected.

**Comment by paulodamaso** [comment]
@g4s8 Closed PR and set impediment in #315.

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#359](https://github.com/zerocracy/datum/issues/359) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@amihaiemil According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only two issues were found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by amihaiemil** [comment]
@ypshenychka confirmed

**Comment by ypshenychka** [comment]
@amihaiemil Thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by ypshenychka** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zerocracy/datum/issues/359#issuecomment-402547433))

@ypshenychka This is what I know about this job in [C3NDPUA8L](https://www.0crat.com/p/C3NDPUA8L), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#359](https://github.com/zerocracy/datum/issues/359) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/C3NDPUA8L?q=%7Bjob%3A%27gh%3Azerocracy%2Fdatum%23359%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by ypshenychka** [comment]
@g4s8 For some reason I got no points and payment here. Could you please do it manually?

**Comment by ypshenychka** [comment]
@g4s8 ping

**Pull request closed without merging**